### PR TITLE
Add `recycle` flag to linear solvers' constructors

### PR DESCRIPTION
**Pull request by jowezarek**
Add a `recycle` flag to the `__init__` of the subclasses of `InverseLinearOperator` defined in `linalg/solvers`.
By default `False`. If `True`, updates the initial guess `x0` to be a copy of the previous solution, without creating a temporary vector.

Previously, this feature could be implicitly used:
Given `A_solv`, an object of one of the solver classes defined in `linalg/solvers`, you could write
`x = A_solv.dot(b, out=A_solv._options["x0"])` instead of simply
`x = A_solv.dot(b)` or equivalently `x = A_solv @ b`.

However, this method has (at least) one major drawback:

**Application in a composition of linear operators**
    Say you have to solve `Sx = b_n` for x at each type step of your simulation, where the right hand side `b_n` changes at each time step, and where S is a composition of linear operators, for example:
    `S = A1 @ A2 @ A_solv @ A3`
    Upon solving for x, `A_solv` is being called in up to `maxiter` (of `S_solv`) iterations, with A being called in up to `maxiter` (of `A_solv`) internal iterations. This number of internal iterations can be drastically decreased upon introducing the `recycle` flag. Implicit usage of this feature is not possible in this case.

A more rare case of the implicit usage not working is the case of a solver being a block of a `BlockLinearOperator`.

### CODE REVIEW

File: psydac/linalg/tests/test_solvers.py [hunk 1]
```diff
@@ -152,10 +174,12 @@ def test_solver_tridiagonal(n, p, dtype, solver, verbose=False):
     #---------------------------------------------------------------------------
     # PYTEST
     #---------------------------------------------------------------------------
-    assert err_norm < tol
-    assert err2_norm < tol
-    assert errt_norm < tol
-    assert errh_norm < tol
+    # The lsmr solver does not consistently produce outputs x whose error ||Ax - b|| is less than tol.
+    if solver != 'lsmr':
```
Comment (yguclu) [inline psydac/linalg/tests/test_solvers.py:178]: What is the problem with the LSMR solver? Does it use a different stopping criteria than the other solvers?

File: psydac/linalg/tests/test_solvers.py [hunk 2]
```diff
@@ -102,31 +102,53 @@ def test_solver_tridiagonal(n, p, dtype, solver, verbose=False):
         print()
 
     #Create the solvers
-    solv  = inverse(A, solver, tol=1e-13, verbose=True)
+    solv  = inverse(A, solver, tol=tol, verbose=False, recycle=True)
```
Comment (yguclu) [inline psydac/linalg/tests/test_solvers.py:105]: If we pass both `x0` and `recycle=True`, will the vector `x0` be modified?

Comment (jowezarek) [reply psydac/linalg/tests/test_solvers.py:105]: Yes. If we don't pass `x0`, we immediately set `x0 = codomain.zeros()`, so there's no possible case where `x0 = None` might change the way any internal functions work.

File: psydac/linalg/tests/test_solvers.py (hunk 1 above)
Comment (jowezarek) [reply psydac/linalg/tests/test_solvers.py:178]: Honestly, I have a lot of trouble understanding LSMR's break condition. We can pass `tol`, `atol` and `btol`, but the break condition reads

```
if istop > 0: break
```

with multiple possible values for `istop`

```
            if itn >= maxiter:istop = 7
            if 1 + test3 <= 1:istop = 6
            if 1 + test2 <= 1:istop = 5
            if 1 + t1 <= 1:istop = 4

            # Allow for tolerances set by the user.

            if test3 <= ctol:istop = 3
            if test2 <= atol:istop = 2
            if test1 <= rtol:istop = 1
```

If we pass `tol`, we set `atol=btol=tol`.
Hence I'd expect `test2` to be the L2 norm `||A*x_n - b||` at iteration n, instead it reads

```
if (normA * normr) != 0:test2 = normar / (normA * normr)
            else:test2 = np.infty
```

and none of the above variables are easy to read.
However, I once printed the break condition and it seemed that the case `istop = 1` was in these test cases always responsible for the early stopping.

**Review by yguclu — APPROVED** [review]

**Pull request merged**